Sources/PostgresNIO/New/PSQLError.swift

```diff
@@ -80,59 +80,55 @@ struct PSQLError: Error {
     }
 }
 
-struct PSQLCastingError: Error {
+struct PSQLCastingError: Error, Equatable {
+    struct Code: Hashable, Error {
+        enum Base {
+            case missingData
+            case typeMismatch
+            case failure
+        }
+
+        var base: Base
+
+        init(_ base: Base) {
+            self.base = base
+        }
+
+        static let missingData = Self.init(.missingData)
+        static let typeMismatch = Self.init(.typeMismatch)
+        static let failure = Self.init(.failure)
```

Comment:
Do I correctly recognize this pattern as being that used by the various `Error` subtypes in `Foundation` (`CocoaError` et al, the ones that use a bunch of underscored protocols to make it all play more nicely with `NSError` bridging)?

Either way, is there any benefit to this pattern beyond that of sidestepping the "is adding enum cases in the absence of library evolution a semver-major break?" issue? Given that Vapor's answer to that question thus far has pretty consistently been "no" aside from a couple of specific cases, if that's the intent here, I'd rather just have an un-nested enum. (Though I'm willing to potentially be argued out of it.)

Reply:
Yes, that’s the adding an enum case is a breaking change is the only reason… I would like to build PostgresNIO in the most resilient way possible, as I consider this lib very foundational. The more headaches we can prevent the better.

And if we can build in a way that allows us to move forward without breaking change, why don’t we take the opportunity?